lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
-            }
-          } else if (WIFSIGNALED(Status)) {
-            // The process exited due to a signal.
-            int Signal = WTERMSIG(Status);
-
-            StringRef ErrorMsg = strsignal(Signal);
-
-            if (Signalled) {
-              TaskFinishedResponse Response =
-                  Signalled(T.getPid(), ErrorMsg, T.getOutput(), T.getErrors(),
-                            T.getContext(), Signal);
-              if (Response == TaskFinishedResponse::StopExecution)
-                // If we have a TaskCrashedCallback, only set SubtaskFailed to
-                // true if the callback returns StopExecution.
-                SubtaskFailed = true;
-            } else {
-              // Since we don't have a TaskCrashedCallback, treat a crashing
-              // subtask as having failed.
-              SubtaskFailed = true;
-            }
-          }
-
-          ExecutingTasks.erase(Pid);
-          FinishedFds.push_back(fd.fd);
-        }
-      } else if (fd.revents & POLLNVAL) {
-        // We passed an invalid fd; this should never happen,
-        // since we always mark fds as finished after calling
-        // Task::finishExecution() (which closes the Task's fd).
-        llvm_unreachable("Asked poll() to watch a closed fd");
-      }
-
-      fd.revents = 0;
-    }
+    Optional<std::vector<int>> FinishedFds =
+        readFromReadyFdsReturningFinishedOnes();
+    if (!FinishedFds)
+      return true;
 
-    // Remove any fds which we've closed from PollFds.
-    for (int fd : FinishedFds) {
-      auto predicate = [&fd] (struct pollfd &i) {
-        return i.fd == fd;
-      };
+    stopPolling(*FinishedFds);
+  }
+  return false;
+}
 
-      auto iter = std::find_if(PollFds.begin(), PollFds.end(), predicate);
-      assert(iter != PollFds.end() && "The finished fd must be in PollFds!");
-      PollFds.erase(iter);
-    }
+bool TaskShepherd::startUpSomeTasks() {
+  while (!TasksToExecute.empty() &&
+         ExecutingTasks.size() < MaxNumberOfParallelTasks) {
+    std::unique_ptr<Task> T = startNextTask();
+    if (!T)
+      return true;
+    if (BeganCallBack)
+      BeganCallBack(T->getPid(), T->getContext());
+    startPollingFdsOfTask(T.get());
+    ExecutingTasks.add(std::move(T));
+  }
+  return false;
+}
+
+void TaskShepherd::startPollingFdsOfTask(const Task *T) {
+  FdsToPoll.push_back({T->getPipe(), POLLIN | POLLPRI | POLLHUP, 0});
+  // We should also poll T->getErrorPipe(), but this introduces timing
+  // issues with shutting down the task after reading getPipe().
+}
+
+TaskShepherd::PollResult TaskShepherd::pollTheFds() {
+  assert(!FdsToPoll.empty() &&
+         "We should only call poll() if we have fds to watch!");
+  int ReadyFdCount = poll(FdsToPoll.data(), FdsToPoll.size(), -1);
+  if (Stats)
+    Stats->getDriverCounters().NumDriverPipePolls++;
+  return ReadyFdCount != -1
+             ? PollResult::noError
+             : errno == EAGAIN || errno == EINTR ? PollResult::softError
```

Comment:
Relying on precedence like this scares me. Can you just use `if`?

Reply:
(I memorized C precedence rules in the early 70's because I liked to keep things terse. :)
I also like using expression-if when there are no side-effects.
But I'm happy to change it.